### PR TITLE
Network Create: Add --ignore flag to support idempotent script

### PR DESCRIPTION
**Pull request by ygalblum**
Add --ignore flag to the command line
NetworkCreate interface - add NetworkCreateOptions parameter Remote API - wrap the Network struct to allow passing additional fields, e.g. ignore Kube - use the IgnoreIfExists flag when creating the default network instead of handling the failure Add e2e tests

Signed-off-by: Ygal Blum <ygal.blum@gmail.com>

#### Does this PR introduce a user-facing change?
Yes

```release-note
Network Create: Add --ignore flag to instruct Podman to not fail when trying to create an already existing Network.
```

### CODE REVIEW

**Comment by ygalblum** [comment]
I needed to create a new structure for the remote APIs. Not sure that I chose the right place for it: https://github.com/containers/podman/pull/16773/files#diff-67310313867eb5378a8f260e7f92ac12683a604b43154ee63c8cd1162a41d889R91

File: pkg/bindings/network/network.go [hunk 1]
```diff
@@ -12,7 +12,7 @@ import (
 )
 
 // Create makes a new CNI network configuration
-func Create(ctx context.Context, network *types.Network) (types.Network, error) {
+func Create(ctx context.Context, network *types.Network, createOptions *types.NetworkCreateOptions) (types.Network, error) {
```
[the remark this answers is not in the document]
Comment (ygalblum) [reply pkg/bindings/network/network.go]: Added a new API. Calls to the old APIs are diverted to the new one with `nil`

File: pkg/api/handlers/libpod/networks.go [hunk 2]
```diff
@@ -25,14 +25,14 @@ func CreateNetwork(w http.ResponseWriter, r *http.Request) {
 	}
 
 	runtime := r.Context().Value(api.RuntimeKey).(*libpod.Runtime)
-	network := types.Network{}
-	if err := json.NewDecoder(r.Body).Decode(&network); err != nil {
+	networkCreateLibpod := entities.NetworkCreateLibpod{}
```
[the remark this answers is not in the document]
Comment (ygalblum) [reply pkg/api/handlers/libpod/networks.go]: Reverted the change to the structure and added a query parameter (setting and getting)

File: test/e2e/network_create_test.go [hunk 3]
```diff
@@ -452,4 +452,33 @@ var _ = Describe("Podman network create", func() {
 		Expect(nc).To(Exit(125))
 		Expect(nc.ErrorToString()).To(Equal("Error: cannot set more ranges than subnets"))
 	})
+
+	It("podman network create same name - fail", func() {
+		name := "subnets-" + stringid.GenerateRandomID()
```
[the remark this answers is not in the document]
Comment (ygalblum) [reply test/e2e/network_create_test.go]: Done

File: pkg/api/handlers/libpod/networks.go [hunk 4]
```diff
@@ -31,8 +31,19 @@ func CreateNetwork(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	query := struct {
+		IgnoreIfExists bool `schema:"ignoreIfExists"`
+	}{
+		IgnoreIfExists: false,
```
Comment (Luap99) [inline pkg/api/handlers/libpod/networks.go:37]: bools are false by default so you do not need to explicitly set this line.

File: pkg/bindings/network/network.go [hunk 5]
```diff
@@ -12,12 +13,21 @@ import (
 )
 
 // Create makes a new CNI network configuration
-func Create(ctx context.Context, network *types.Network) (types.Network, error) {
+func Create(ctx context.Context, network *types.Network, createOptions *NetworkCreateOptions) (types.Network, error) {
```
Comment (Luap99) [inline pkg/bindings/network/network.go:16]: looks like you changed the wrong place, pkg/bindings is stable so this is a breaking change
pkg/domain is not stable so you can break it there and do not need to keep the old function

**Review by Luap99 — COMMENTED** [review]
You also need to update the podman-network-create man page to include the new option.

File: pkg/api/handlers/libpod/networks.go (hunk 4 above)
Comment (ygalblum) [reply pkg/api/handlers/libpod/networks.go]: Right, fixed

File: pkg/bindings/network/network.go (hunk 5 above)
Comment (ygalblum) [reply pkg/bindings/network/network.go]: I guess I didn't notice where the comment was and assumed I knew. Fixed

File: docs/source/markdown/podman-network-create.1.md [hunk 6]
```diff
@@ -39,6 +39,9 @@ Define a gateway for the subnet. If you want to provide a gateway address, you m
 *subnet* option. Can be specified multiple times.
 The argument order of the **--subnet**, **--gateway** and **--ip-range** options must match.
 
+#### **--ignore**
+Ignore the create request if a network with the same name already exists instead of failing
```
Comment (mheon) [inline docs/source/markdown/podman-network-create.1.md:43]: Probably worth mentioning that only the name is checked, so `podman network create --driver macvlan testnet` will still exit cleanly if `testnet` exists but isn't macvlan, or similar.

**Comment by ygalblum** [comment]
@Luap99 I've fixed the comments. In addition, can you please restart the failed, test? It doesn't look like it has anything to do with this change.

File: docs/source/markdown/podman-network-create.1.md (hunk 6 above)
Comment (ygalblum) [reply docs/source/markdown/podman-network-create.1.md]: Please see the line I've added

File: pkg/bindings/network/network.go [hunk 7]
```diff
@@ -13,11 +14,24 @@ import (
 
 // Create makes a new CNI network configuration
 func Create(ctx context.Context, network *types.Network) (types.Network, error) {
+	return CreateWithOptions(ctx, network, nil)
+}
+
+func CreateWithOptions(ctx context.Context, network *types.Network, extraCreateOptions *ExtraCreateOptions) (types.Network, error) {
```
Comment (Luap99) [inline pkg/bindings/network/network.go:20]: I think naming it just `CreateOptions` sounds better but I am also ok with this one.

**Review by Luap99 — COMMENTED** [review]
LGTM

File: pkg/bindings/network/network.go (hunk 7 above)
Comment (ygalblum) [reply pkg/bindings/network/network.go:20]: The code already has `CreateOptions` struct (this is the reason I named the new struct `ExtraCreateOptions`). So, I think the `With` does help here

Comment (Luap99) [reply pkg/bindings/network/network.go:20]: Oh no, looks like I forgot to remove the old type when I broke the network code with the v4.0 bump.
Make sense in this case.

Comment (ygalblum) [reply pkg/bindings/network/network.go:20]: OK, resolving then

**Comment by ygalblum** [comment]
@mheon can you please take a look?

File: docs/source/markdown/podman-network-create.1.md (hunk 6 above)
Comment (mheon) [reply docs/source/markdown/podman-network-create.1.md]: Looks fine

**Pull request merged**